src/include/sof/debug/panic.h

```diff
@@ -32,14 +32,26 @@ void panic_dump(uint32_t p, struct sof_ipc_panic_info *panic_info,
 
 #ifdef __ZEPHYR__
 #include <kernel.h>
+#include <sys/__assert.h>
+
 #define panic(x) k_panic()
 
+#define verify_assert(cond) \
+	do {	       \
+		if (!(cond)) \
+			k_oops(); \
+	} while (0)
+
+#if __ASSERT_ON
 #define assert(x) \
-	do {			\
-		if (!(x))	\
-			k_oops();\
+	do { \
+		__ASSERT_NO_MSG(x); \
 	} while (0)
-#else
+#else /* !__ASSERT_ON */
+#define assert(x) (void)(x)
+#endif
```

Comment:
You're saying that currently SOF's usage of `assert()` has side-effects, so if we compile it out, unexpected things might happen, right? So, to fix this we first have to eliminate those potential side effects, and just disabling `assert()` could break things. But here you make `assert()` depend on `__ASSERT_ON` which isn't defined in our current builds, so this patch already disables `assert()`? Am I missing something?

Reply:
Ack @lyakh , I'll add more comments. Intention is definitely to keep bisect working. In short with this patch, the assert expression is always evaluated:
 case A) If you have Zephyr's __ASSERTs enabled (CONFIG_ASSERT=y and friends), we call "__ASSERT_NO_MSG(x)" and code is generated to evaluate (x).
  case B) If you have __ASSERT==0 (e.g. no code is produced for "__ASSERT_NO_MSG(x)"), we emit the expression to compiler output directly with "(void)(x)".

So e.g. "assert(memcpy())" always keeps the memcpy() in compiler output.